File: framework/java/src/main/java/soda/unittest/StructTester.java
package soda.unittest;

import soda.unittest.conv.ConverterFactory;
import soda.unittest.conv.ObjectConverter;

import java.util.ArrayList;
import java.util.List;

public class StructTester {

    private Class<?> structClass;

    public StructTester(Class<?> structClass) {
        this.structClass = structClass;
    }

    static TestWork createTestWork(Class<?> structClass) {
        return new TestWork(new StructTester(structClass), "test");
    }

    public Object test(List<String> operations, List<Object> parameters) {
        return Utils.wrapEx(() -> {
            var ctor = structClass.getConstructors()[0];
            ctor.setAccessible(true);
            var cparams = Utils.parseArguments(ctor.getParameterTypes(), Utils.cast(parameters.get(0)));
            var obj = ctor.newInstance(cparams);
            var res = new ArrayList<>();
            res.add(null);
            for (int i = 1; i < parameters.size(); ++i) {
                var method = Utils.findMethod(structClass, operations.get(i));
                method.setAccessible(true);
                var params = Utils.parseArguments(method.getGenericParameterTypes(), Utils.cast(parameters.get(i)));
                var r = method.invoke(obj, params);
                ObjectConverter<Object, Object> conv = ConverterFactory.create(method.getGenericReturnType());
                res.add(conv.toJsonSerializable(r));
            }
            return res;
        });
    }

}

File: framework/kotlin/src/main/kotlin/soda/kotlin/unittest/StructTester.kt
package soda.kotlin.unittest

import kotlinx.serialization.json.*
import soda.kotlin.unittest.conv.ConverterFactory
import kotlin.reflect.KClass
import kotlin.reflect.typeOf
import kotlin.reflect.full.memberFunctions
import kotlin.reflect.full.primaryConstructor

class StructTester(private val structClass: KClass<*>) {
    fun test(operations: List<String>, parameters: JsonElement): JsonElement {
        val ctor = structClass.primaryConstructor ?:
            throw NoSuchMethodError("no constructor for class $structClass")
        val ctorParamTypes = ctor.parameters.map { it.type }
        val ctorParams = Utils.parseArguments(
                ctorParamTypes, parameters.jsonArray[0])
        val obj = ctor.call(*ctorParams.toTypedArray())
        val res = mutableListOf<JsonElement>()
        res.add(JsonNull)

        val methodMap = structClass.memberFunctions.associateBy { it.name }
        for (i in 1 until parameters.jsonArray.size) {
            val methodName = operations[i]
            val method = methodMap[methodName] ?:
                throw NoSuchMethodError("no member function $methodName for class $structClass")
            val paramTypes = method.parameters.drop(1).map { it.type }
            val params = Utils.parseArguments(paramTypes, parameters.jsonArray[i])
            val r = method.call(obj, *params.toTypedArray())
            val retType = method.returnType
            if (retType.toString() != typeOf<Unit>().toString()) {
                val conv = ConverterFactory.create<Any?>(retType)
                res.add(conv.toJsonSerializable(r))
            } else {
                res.add(JsonNull)
            }
        }
        return JsonArray(res)
    }
}
